libraries/Microsoft.Bot.Builder/TokenResolver.cs

```diff
@@ -0,0 +1,134 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Security.Claims;
+using System.Security.Principal;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder
+{
+    internal class TokenResolver
+    {
+        private const int PollingInterval = 1000;   // Poll for token every 1 second.
+
+        public static void CheckForOAuthCards(BotFrameworkAdapter adapter, ITurnContext turnContext, Activity activity, CancellationToken cancellationToken)
+        {
+            if (activity?.Attachments != null)
+            {
+                foreach (var attachment in activity?.Attachments?.Where(a => a.ContentType == OAuthCard.ContentType))
+                {
+                    OAuthCard oauthCard = attachment.Content as OAuthCard;
+                    if (oauthCard != null && !string.IsNullOrEmpty(oauthCard.ConnectionName))
+                    {
+                        // Poll as a background task
+                        Task.Run(() => PollForTokenAsync(adapter, turnContext, activity, oauthCard.ConnectionName, cancellationToken));
```

Comment:
Have you evaluated the advantages of using this vs `HostingEnvironment.QueueBackgroundWorkItem`? SEe here for more details: https://www.hanselman.com/blog/HowToRunBackgroundTasksInASPNET.aspx.

Reply:
Well, as this is doing a small amount of a work at an interval, the blog says: 
"If your ASP.NET application just needs one background task to runs an a basic scheduled interval, than perhaps you just need the basics of WebBackgrounder.", which is Task.Run. 

Also, when I reached out to the ASP.NET team a few weeks ago to talk to them about thread pools, they said with ASP.NET Core there isn't a a separate thread pool anymore and that Task.Run was functionally equivalent to more complex mechanisms (except when you really want your own task scheduler). This could be Core or NetFx I guess, but still it's just running at an interval.

That's a good article though :-)